### PR TITLE
event: support socket sources on Windows

### PR DESCRIPTION
**Pull request by adierking**
Provide an initial implementation of Windows socket sources using
WSAEventSelect(). This is sufficient for the dispatch_io_net test to
pass on Windows.

### CODE REVIEW

**Comment by adierking** [comment]
cc @compnerd @ktopley-apple

File: src/event/event_windows.c [hunk 1]
```diff
@@ -146,17 +159,12 @@ _dispatch_muxnote_create(dispatch_unote_t du,
...
-			// anonymous pipes.
-			if (!GetNamedPipeInfo(handle, NULL, NULL, NULL, NULL)) {
-				// We'll get ERROR_ACCESS_DENIED for outbound pipes.
-				if (GetLastError() != ERROR_ACCESS_DENIED) {
-					// The file is probably a socket.
-					WIN_PORT_ERROR();
-				}
+			// anonymous pipe.
+			if (_dispatch_handle_is_socket(handle)) {
+				dmn->dmn_handle_type = DISPATCH_MUXNOTE_HANDLE_TYPE_SOCKET;
```
[the remark this answers is not in the document]
Comment (adierking) [reply src/event/event_windows.c]: No, there shouldn't be any new cases. I can make this a ternary.

**Comment by adierking** [comment]
>Should libdispatch invoke WSAStartup somewhere perhaps? Although, I suppose that if the user is creating a socket, they will have initialised winsock and everything should be fine. libdispatch will never construct a socket on its own (unless it is in a test, where we can explicitly call WSAStartup.

Right, since libdispatch only ever consumes sockets, I think it would be best to let clients initialize WinSock on their own. We don't have any means of knowing whether a program wants to use sockets otherwise, so this might lead to WinSock being unnecessarily initialized for programs that don't need to do any networking.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Pull request merged**